src/transcript.rs

```diff
@@ -2,125 +2,221 @@
 //! transcripts.
 
 use ff::Field;
-use std::marker::PhantomData;
 use std::ops::Deref;
 
 use crate::arithmetic::{CurveAffine, FieldExt};
 
-/// This is a generic interface for a sponge function that can be used for
-/// Fiat-Shamir transformations.
-pub trait Hasher<F: FieldExt>: Clone + Send + Sync + 'static {
-    /// Initialize the sponge with some key.
-    fn init(key: F) -> Self;
-    /// Absorb a field element into the sponge.
-    fn absorb(&mut self, value: F);
-    /// Square a field element out of the sponge.
-    fn squeeze(&mut self) -> F;
+use std::io::{self, Read, Write};
+use std::marker::PhantomData;
+
+/// Generic transcript view (from either the prover or verifier's perspective)
+pub trait Transcript<C: CurveAffine> {
+    /// Squeeze a challenge (in the base field) from the transcript.
+    fn squeeze_challenge(&mut self) -> C::Base;
+
+    /// Writing the point to the transcript without writing it to the proof,
+    /// treating it as a common input.
+    fn common_point(&mut self, point: C) -> io::Result<()>;
+}
+
+/// Transcript view from the perspective of a verifier that has access to an
+/// input stream of data from the prover to the verifier.
+pub trait TranscriptRead<R: Read, C: CurveAffine>: Transcript<C> {
```

Comment:
I think that the `R` type parameter is unnecessary and can be removed, especially after removing the need for transcript forking.